src/fmt_plain_writer.rs

```diff
@@ -0,0 +1,153 @@
+        match self.current_newline_stretch {
+            None | Some(0) => false,
+            Some(_) => true,
+        }
+    }
+
+    pub fn take_underlying(self) -> W {
+        self.underlying
+    }
+
+    fn flush_newlines(&mut self) -> std::io::Result<()> {
+        if let Some(newlines) = self.current_newline_stretch {
+            for _ in 0..min(newlines, self.max_newlines) {
+                writeln!(self.underlying)?;
+            }
+        }
+        // Set the current stretch to 0 -- not to None, since we want to note here that we've written something!
+        self.current_newline_stretch = Some(0);
+        Ok(())
+    }
+
+    fn increment_newline_stretch(&mut self) {
+        self.current_newline_stretch = Some(match self.current_newline_stretch {
+            None => 0,
+            Some(n) => n + 1,
+        });
+    }
+}
+
+impl<W: Write> Write for NewlineCollapser<W> {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut wrote = 0;
+        let mut remaining = buf;
+        while !remaining.is_empty() {
+            match memchr::memchr(b'\n', remaining) {
+                None => {
+                    // No newline found, and we know there's at least one byte due to the `while` condition.
+                    // So: (1) write `remaining` to the underlying, (2) set just_wrote_newline = false (since we just
+                    // wrote at least one byte, and no newlines), and (3) break, since we just wrote all of remaining.
+                    self.flush_newlines()?;
+                    wrote += self.underlying.write(remaining)?;
+                    self.current_newline_stretch = Some(0);
+                    break;
+                }
+                Some(0) => {
+                    // First byte is a newline. Increment the current stretch, and that's it.
+                    // This case is the whole purpose of this struct: the `else` is what does the newline collapsing.
+                    self.increment_newline_stretch();
+                    wrote += 1; // We did process this byte, even if we haven't actually written it out yet
+                    remaining = &remaining[1..];
+                }
+                Some(n) => {
+                    // The first byte isn't a newline, so even if we had just written a newline previously, we can
+                    // always just write out that first char. Keep writing until n. If we wrote n bytes, then we wrote
+                    // the newline; otherwise, we didn't.
+                    self.flush_newlines()?;
+                    let underlying_wrote_n = self.underlying.write(&remaining[..n])?;
+                    wrote += underlying_wrote_n;
+                    if underlying_wrote_n == n {
+                        self.increment_newline_stretch();
+                        wrote += 1;
+                    }
+                    remaining = &remaining[underlying_wrote_n + 1..];
+                }
+            }
+        }
```

Comment:
@coderabbitai Do you see a way to simplify this? It feels like it'd be more complicated than I'd expect, but maybe it really is this complex a problem.

Reply:
> Eliminate the Option type - Initialize current_newline_stretch to 0 instead of None. The None adds complexity without much benefit

This is needed in order to remove leading newlines.